cpp/include/cudf/detail/iterator.cuh

```diff
@@ -134,6 +232,58 @@ auto make_null_replacement_iterator(column_device_view const& column,
       null_replaced_value_accessor<Element>{column, null_replacement});
 }
 
+/**
+ * @brief Constructs a pair iterator over a column's values that replaces null
+ * elements with a specified value and its validity.
+ *
+ * Dereferencing the returned iterator for element `i` will return `pair(column[i], true)`
+ * if it is valid, or `pair(null_replacement, false)` if it is null.
+ * This iterator is only allowed for nullable columns.
+ *
+ * @throws `cudf::logic_error` if the column is not nullable.
+ * @throws `cudf::logic_error` if column datatype and Element type mismatch.
+ *
+ * @tparam Element The type of elements in the column
+ * @param column The column to iterate
+ * @param null_replacement The value to return for null elements
+ * @return auto Iterator that returns valid column elements, or a null
+ * replacement value for null elements and validity of the element in a pair
+ */
+template <typename Element>
+auto make_null_replacement_pair_iterator(column_device_view const& column,
+                                    Element const null_replacement = Element{0})
+{
+  return thrust::make_transform_iterator(
+      thrust::counting_iterator<cudf::size_type>{0},
+      null_replaced_pair_accessor<Element>{column, null_replacement});
+}
+
+/**
+ * @brief Constructs a pair iterator over a column's values and its validity.
+ *
+ * Dereferencing the returned iterator for element `i` will return
+ * `pair(column[i], validity)`.
+ * `validity` is `true` if `has_nulls=false`.
+ * `validity` is validity of the element at `i` if `has_nulls=true` and the
+ * column is nullable.
+ *
+ * @throws `cudf::logic_error` if the column is nullable.
+ * @throws `cudf::logic_error` if column datatype and Element type mismatch.
+ *
+ * @tparam Element The type of elements in the column
+ * @tparam has_nulls boolean indicating to treat the column is nullable
+ * @param column The column to iterate
+ * @return auto Iterator that returns valid column elements, and validity of the
+ * element in a pair
+ */
+template <typename Element, bool has_nulls=false> 
+auto make_pair_iterator(column_device_view const& column)
```

Comment:
We also need one of these that works for `cudf::scalar`.

Reply:
`cudf::scalar` contains a single value. Do we need an _iterator_ for `cudf::scalar`?

Reply:
Yes, several places have needed this functionality (e.g., https://github.com/rapidsai/cudf/pull/3471). Just make a `thrust::constant_iterator` around the scalar's value.

Reply:
added.